backend/benefit/applications/tests/test_benefit_aggregation.py

```diff
@@ -0,0 +1,363 @@
+                    "2020-07-01",
+                    "2020-12-31",
+                ),
+            ],
+            BenefitType.SALARY_BENEFIT,
+            NO_APPRENTICESHIP,
+            400,
+            12,
+            0,
+        ),  # 24 months not elapsed
+        (
+            [
+                (
+                    Application,
+                    "2018-01-01",
+                    "2018-06-30",
+                ),
+                (
+                    Application,
+                    "2018-07-01",
+                    "2018-12-31",
+                ),
+            ],
+            BenefitType.SALARY_BENEFIT,
+            NO_APPRENTICESHIP,
+            200,
+            0,
+            12,
+        ),  # 24 months elapsed
+        (
+            [
+                (
+                    Application,
+                    "2017-07-01",
+                    "2018-06-30",
+                ),
+                (
+                    Application,
+                    "2020-07-01",
+                    "2020-12-31",
+                ),
+            ],
+            BenefitType.SALARY_BENEFIT,
+            NO_APPRENTICESHIP,
+            200,
+            6,
+            6,
+        ),
+        # 24-month gap in past benefits so the benefit from 2017-2018 is not included and application is valid
+        (
+            [
+                (
+                    PreviousBenefit,
+                    "2018-01-01",
+                    "2018-03-31",
+                ),
+                (
+                    Application,
+                    "2019-01-01",
+                    "2019-03-31",
+                ),
+                (
+                    Application,
+                    "2020-01-01",
+                    "2020-03-31",
+                ),
+            ],
+            BenefitType.SALARY_BENEFIT,
+            NO_APPRENTICESHIP,
+            400,
+            9,
+            3,
+        ),
+        # several previously granted benefits that don't have a 24-month gap
+    ],
+)
+def test_application_with_previously_accepted_applications_and_previous_benefits(
+    api_client,
+    application,
+    previous_benefits,
+    benefit_type,
+    apprenticeship_program,
+    months_used,
+    months_remaining,
+    expected_result,
+):
+    for class_name, previous_start_date, previous_end_date in previous_benefits:
+        # previous, already granted benefits for the same employee+company
+        if class_name == Application:
+            decided_application = DecidedApplicationFactory()
+            # the type of the previous benefit and the apprenticeship_program value do not matter when
+            # calculating the remaining available benefit time (Teams discussion)
+            decided_application.benefit_type = random.choice(
+                [BenefitType.SALARY_BENEFIT, BenefitType.EMPLOYMENT_BENEFIT]
+            )
+            decided_application.apprenticeship_program = random.choice([True, False])
```

Comment:
Does this produce reproducible tests in the review environment?
I can see that in backend/benefit/common/tests/conftest.py there's
```
@pytest.fixture(autouse=True)
def setup_test_environment(settings):
    factory.random.reseed_random("777")
    with freeze_time("2021-06-04"):
        yield
```
which makes the tests at least under backend/benefit/common/tests reproducible
but it doesn't affect the tests in this test's directory, right?

It'd probably be better to make the random choices always be the same random
choices when the tests are run so that the tests don't randomly fail and succeed
if some values happen to be corner cases. What do you think?

Reply:
fixed

Reply:
Sorry, I gave you false information. I just had a talk with @quyenlq in another PR on this same subject, see the following comments: https://github.com/City-of-Helsinki/yjdh/pull/545#discussion_r771393436, https://github.com/City-of-Helsinki/yjdh/pull/545#discussion_r771400958, https://github.com/City-of-Helsinki/yjdh/pull/545#discussion_r772105156.

Now it actually runs the `backend/benefit/common/tests/conftest.py`'s `setup_test_environment` for each parametrization of `test_application_with_previously_accepted_applications_and_previous_benefits` so
```python
decided_application.benefit_type = random.choice(
                [BenefitType.SALARY_BENEFIT, BenefitType.EMPLOYMENT_BENEFIT]
            )
```
will always give the same value.

I thought this would've generated the same random values for each parametrization of the function `test_application_with_previously_accepted_applications_and_previous_benefits` i.e.
 - parametrization 0 → random value X₀
 - parametrization 1 → random value X₁
 - …
 - parametrization N → random value Xₙ

So that X₀, X₁, …, Xₙ would be random but stay the same between different pytest test runs.
But it became apparent to me this morning that this is not so as I doublechecked my own code in another PR.